### PR TITLE
Generate checluster references tables

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

> Read our [Contribution guide](https://github.com/eclipse/che-docs/blob/master/CONTRIBUTING.adoc) before submitting a PR.

### What does this PR do?
Add scripts to sync doc and checluster crd

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/RHDEVDOCS-2216

### PR Checklist

As the author of this Pull Request I made sure that:

- [ ] `vale` has been run successfully against the PR branch
- [x] Link checker has been run successfully against the PR branch
- [ ] Documentation describes a scenario that is already covered by QE tests, otherwise an issue has been created and acknowledged by Che QE team
- [x] Changed article references are updated where they are used (or a redirect has been set up on the docs side):
    - [x] Dashboard [branding.json](https://github.com/eclipse/che-dashboard/blob/master/src/components/branding/branding.json)
    - [x] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)

### CODE REVIEW

**Review by MichalMaler — APPROVED** [review]

**Comment by MichalMaler** [comment]
LGTM. We are waiting for @l0rd approval now.

**Comment by tolusha** [comment]
gen script requires `yq` installed since it will be run on Jenkins

**Comment by MichalMaler** [comment]
> gen script requires `yq` installed since it will be run on Jenkins

We can add:

.Prerequisites
* The `yq` tool is available. See link:https://pypi.org/project/yq/[yq homepage].

**Comment by themr0c** [comment]
The che-docs container has `yq` inside.

**Comment by themr0c** [comment]
Can you confirm it runs on che.openshift.io?

**Comment by tolusha** [comment]
I've just checked:
1. Created a workspace from the devfile [1]
2. Run `yq` command in `che-docs` containers.
yes, it exists there.

[1] https://raw.githubusercontent.com/eclipse/che-docs/master/devfile.yaml

**Comment by themr0c** [comment]
@MichalMaler the requirement is to build the docs, not to the user-facing docs.

**Review by themr0c — CHANGES_REQUESTED** [review]
Can you fix the shellcheck errors:

shellcheck tools/checluster_docs_gen.sh 

In tools/checluster_docs_gen.sh line 17:
TABLE_HEADER="$NEWLINE[cols=\"2,5\", options="header"]$NEWLINE:=== $NEWLINE Property: Description $NEWLINE"
              ^-- SC1087: Use braces when expanding arrays, e.g. ${array[idx]} (or ${var}[.. to quiet).
                                              ^----^ SC2140: Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"?


In tools/checluster_docs_gen.sh line 19:
PARENT_PATH=$( cd "$(dirname "${BASH_SOURCE[0]}")/.." ; pwd -P )
               ^-- SC2164: Use 'cd ... || exit' or 'cd ... || return' in case cd fails.

Did you mean: 
PARENT_PATH=$( cd "$(dirname "${BASH_SOURCE[0]}")/.." || exit ; pwd -P )


In tools/checluster_docs_gen.sh line 27:
  if [ $? -ne 0 ]; then
       ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.


In tools/checluster_docs_gen.sh line 41:
  if [ $? -ne 0 ]; then
       ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.


In tools/checluster_docs_gen.sh line 59:
  if [ $? -ne 0 ]; then
       ^-- SC2181: Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.


In tools/checluster_docs_gen.sh line 88:
    local section=$(echo "$RAW_CONTENT" | yq -M '.spec.validation.openAPIV3Schema.properties.status')
          ^-----^ SC2155: Declare and assign separately to avoid masking return values.


In tools/checluster_docs_gen.sh line 90:
    local section=$(echo "$RAW_CONTENT" | yq -M '.spec.validation.openAPIV3Schema.properties.spec.properties.'$sectionName)
          ^-----^ SC2155: Declare and assign separately to avoid masking return values.
                                                                                                              ^----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
    local section=$(echo "$RAW_CONTENT" | yq -M '.spec.validation.openAPIV3Schema.properties.spec.properties.'"$sectionName")


In tools/checluster_docs_gen.sh line 94:
    $(echo "$section" | yq -M '.properties | keys[]' )
    ^-- SC2207: Prefer mapfile or read -a to split command output (or quote to avoid splitting).


In tools/checluster_docs_gen.sh line 103:
    description=$(echo "$section" | yq -M '.properties.'$prop'.description')
                                                        ^---^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
    description=$(echo "$section" | yq -M '.properties.'"$prop"'.description')

For more information:
  https://www.shellcheck.net/wiki/SC1087 -- Use braces when expanding arrays,...
  https://www.shellcheck.net/wiki/SC2140 -- Word is of the form "A"B"C" (B in...
  https://www.shellcheck.net/wiki/SC2155 -- Declare and assign separately to ...

**Comment by tolusha** [comment]
@themr0c 
Sorry, to be honest I don't think that they are relevant. Where did they come from ?

**Comment by l0rd** [comment]
> @themr0c
> Sorry, to be honest I don't think that they are relevant. Where did they come from ?

@tolusha [shellcheck](https://github.com/koalaman/shellcheck) is the most widespread linter for bash scripts (22K stars) and it's [common](https://github.com/eclipse/che/pull/17865) [to run](https://github.com/eclipse/che/pull/13542#discussion_r299337657) [our scripts](https://github.com/eclipse/che/pull/12455/commits/b818fccea3496f85579ad1affaf880674bb718f1) against it.

**Comment by tolusha** [comment]
@themr0c 
Fixed

**Comment by MichalMaler** [comment]
@themr0c Fabrice, I think we are good to go here :) 
Good job @tolusha  and thank you, @themr0c for setting up the linter.

**Comment by tolusha** [comment]
```bash
$ shellcheck checluster_docs_gen.sh 
$ 
```

no issues with spellchecker anymore ...

**Comment by themr0c** [comment]
So now, what can we do for the vale errors?

**Comment by MichalMaler** [comment]
let me take a look

**Comment by themr0c** [comment]
To fix vale errors:

For the static file, `ref_checluster-custom-resource-fields-reference.adoc`, this is straightforward: fix all vale errors in place.

For the generated file:
* Long term: fix the upstream of the file.
* Short term: substitutions, as we do with `environment_docs_gen.sh`

File: tools/checluster_docs_gen.sh
Comment (themr0c) [inline tools/checluster_docs_gen.sh]: As we have `yq` as a dependency, you can use it here too.

**Comment by tolusha** [comment]
Too much of vale errors, that's not possible to fix at script using substitution. 
I suggest the following:
1. Fix build error and merge.
2. Doc team reviews CRD file to fix descriptions there.
3. Regenerate the doc with a new CRD

File: tools/checluster_docs_gen.sh
Comment (themr0c) [inline tools/checluster_docs_gen.sh]: Use `yq` here too?

File: tools/checluster_docs_gen.sh
Comment (themr0c) [inline tools/checluster_docs_gen.sh]:
```suggestion
# Copyright (c) 2021 Red Hat, Inc.
```

**Comment by MichalMaler** [comment]
> To fix vale errors:
> 
> For the static file, `ref_checluster-custom-resource-fields-reference.adoc`, this is straightforward: fix all vale errors in place.
> 
> For the generated file:
> 
> * Long term: fix the upstream of the file.
> * Short term: substitutions, as we do with `environment_docs_gen.sh`

Here is my Vale fix in the file, but I am concern that this was useless work since I was working on this instead of monitoring your conversation.
https://github.com/tolusha/che-docs/pull/4

@tolusha Please, can you use it to fix the Origin?

**Comment by themr0c** [comment]
Some hints to move forward in testing and validating this PR:

To run the script in `che-docs`:

* Backport the file in che-operator to the `7.25` release. This is the only option to validate properly the PR in che-docs.
* Same thing for the `7.24` release.
* To test and validate beforehand (don't commit in the PR): change the `prod-ver` value to `master` in `antora-playbook.yml`

To run `vale` in `che-operator` repository which has no .vale.ini

* Point vale to the che-docs config file  
```vale --config='some/file/path/.vale.ini'```

**Comment by tolusha** [comment]
@themr0c 
Thank you. I will handle that.

But there is an open question: Jenkins build failed.
I think docker image contains incorrect yq package.
It should be https://github.com/kislyuk/yq instead of  http://mikefarah.github.io/yq/

**Comment by themr0c** [comment]
So 
```
pip3 install yq
```
Instead of 

```
apk add yq
```
?

**Comment by themr0c** [comment]
@tolusha Waiting for validation of https://github.com/eclipse/che-docs/pull/1822 before pushing the new che-docs container to quay.

**Comment by themr0c** [comment]
@tolusha the correct version of `yq` is now available in the che-docs container.

**Review by themr0c — APPROVED** [review]

**Comment by themr0c** [comment]
As we still have lots of language errors, if we merge now, we'll need a follow-up task to:
- fix the language in the source of these generated docs,
- then adapt the script to execute substitutions for what can't be changed upstream (prod-short attribute instead of Che).

**Comment by tolusha** [comment]
@themr0c 
We agreed with @MichalMaler to review the source one more time.

**Comment by MichalMaler** [comment]
I will do it when I take care of RNaKI. Hope I will start with this task today.

**Comment by tolusha** [comment]
I've just updated PR based on the latest changes from master branch of che-operator to see if we have any vale errors.

**Comment by tolusha** [comment]
I've update checluster doc based on master che-operator branch. So, we don't have value errors anymore.
But jenkins still fails.

**Comment by tolusha** [comment]
NB: If PR is merged then checluster references tables will be regenerated based on 7.25 version of che-operator (from antora-playbook.yaml) which doesn't contain all vale fixes.

**Review by MichalMaler — APPROVED** [review]

**Pull request merged**
